### ISSUE TITLE
Better error handling

### ISSUE THREAD

**Issue by kobermeit** [altangent/ccxws#57]
I've had CCXWS running on multiple exchanges and assets for a while now, and recently noticed that some exchanges (Binance in this case) will throw an error and cause my script to crash, even with a try catch around it. Below is the snippet I receive. I am trying to add logic to my code so I can catch this error and handle it myself and believe this is a bug. If it isn't, please let me know how to handle it as there doesn't appear to be any documentation and what I've tried so far hasn't worked (try catch and websocket.on('error')).

`info: watcher initiating reconnection
info: reconnecting
info: closing connection to wss://stream.binance.com:9443/stream?streams=bnbusdt@aggTrade
info: attempting connection
info: connecting to wss://stream.binance.com:9443/stream?streams=bnbusdt@aggTrade
events.js:183
      throw er; // Unhandled 'error' event
      ^

RangeError: Invalid WebSocket frame: invalid status code 1005
    at Receiver.controlMessage (/var/www/html/discord-node/node_modules/ccxws/node_modules/ws/lib/receiver.js:441:18)
    at Receiver.getData (/var/www/html/discord-node/node_modules/ccxws/node_modules/ws/lib/receiver.js:329:42)
    at Receiver.startLoop (/var/www/html/discord-node/node_modules/ccxws/node_modules/ws/lib/receiver.js:133:22)
    at Receiver._write (/var/www/html/discord-node/node_modules/ccxws/node_modules/ws/lib/receiver.js:69:10)
    at doWrite (_stream_writable.js:397:12)
    at writeOrBuffer (_stream_writable.js:383:5)
    at Receiver.Writable.write (_stream_writable.js:290:11)
    at TLSSocket.socketOnData (/var/www/html/discord-node/node_modules/ccxws/node_modules/ws/lib/websocket.js:795:35)
    at emitOne (events.js:116:13)
    at TLSSocket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at TLSSocket.Readable.push (_stream_readable.js:208:10)
    at TLSWrap.onread (net.js:607:20)
error: Forever detected script exited with code: 1
error: Script restart attempt #179`

**Comment by bmancini55** [comment]
Interesting, we haven't encountered this before. What OS and version of node is this happening on?  

`SmartWss` used by Binance is handling error events emitted by the underlying socket: https://github.com/altangent/ccxws/blob/master/src/smart-wss.js#L69.  So I don't think that is the problem.

There might be an update in the underlying `ws` package that could impact this.

**Comment by kobermeit** [comment]
I've tried it on 2 different OSes - Mac with Node 11.10 and my VPS which is Ubuntu Node 8.11.8.

The problem has gone away now, I have a feeling it was due to Binance rate limits or the server just throwing something. I ended up disabled binance and then started it the next day and it worked fine, then I spam restarted my script and it happened again, so really sounds like rate limits. Other exchanges like bitmex returned a 429, binance was the only one doing this 1005 status code.

My current CCXWS version is 0.21.3

**Comment by bmancini55** [comment]
Thanks for the info! It certainly sounds like a limit issue.

The log you posted above shows that it's performing a reconnection event that was initiated by the watcher. If you're only running a single market, you may want to increase the watcher's timeout to prevent reconnection churn.  The default reconnects when no data is seen in a 30s window, which would be overly aggressive for a slower moving market. 

```javascript
const {Watcher} = require('ccxws');
const binance = new ccxws.binance();
binance._watcher = new Watcher(binance, 15 * 60 * 1000); // change to 15 minutes
```

I'm also going to make a change to the Binance client that resets the watcher when the `ping` event is received by the client. This should fix reconnection churn without having to override the watcher.

**Comment by kobermeit** [comment]
Oh thanks, I wasn't aware you could do this! I was originally tracking more popular markets like BTCUSDT and ETHUSDT but its possible I had disabled them too. I'll give this a go as it makes perfect sense.

Going to close this. Cheers!

**Comment by kobermeit** [comment]
**UPDATE: ** Since adding this code an hour ago, the websocket hasn't disconnected once (when previously it was doing it every minute and the 1005 was appearing). I've tried with 20 assets and no errors, working perfectly. Thanks for your help!

**Comment by bmancini55** [comment]
@kobermeit excellent!  

We just pushed a new version of CCXWS that increased the watcher timeout for Binance and allows use of the `reconnectTimeoutMs` constructor argument if you need to customize it (which is a bit simpler than the workaround I previously posted).  No need to update to update to the new version unless you want to as the workaround will work just fine.

More info: https://github.com/altangent/ccxws/releases/tag/v0.23.0

**Comment by evan-coygo** [comment]
@bmancini55  is that `reconnectTimeoutMs` argument still available? I don't see it anywhere in the code or docs

**Comment by bmancini55** [comment]
Apologies, the property was `reconnectIntervalMs` as defined here:
https://github.com/altangent/ccxws/blob/master/src/exchanges/binance-client.js#L15

The underlying cause of the unhandled exceptions was also recently addressed in this commit: https://github.com/altangent/ccxws/commit/a5fd310d2e4d93df0f8aeb1d85b62f8ffb1ee2c7

It is available in v0.26.1+
